lib/Transforms/Scalar/DxilLoopUnroll.cpp

```diff
@@ -0,0 +1,833 @@
+           !isa<PHINode>(I->user_back())))
+        continue;
+
+      Changed |= processInstruction(Body, *L, *I, DT, ExitBlocks, PredCache, LI);
+    }
+  }
+
+  return Changed;
+}
+
+static bool DoesBlockNeedUnroll(BasicBlock *BB, const SetVector<Instruction *> &LoopDependencies) {
+  for (Instruction &I : *BB) {
+    if (GetElementPtrInst *GEP = dyn_cast<GetElementPtrInst>(&I)) {
+      Type *EltType = GEP->getType()->getPointerElementType();
+      if (!hlsl::dxilutil::IsHLSLObjectType(EltType))
+        continue;
+      for (auto IdxIt = GEP->idx_begin(); IdxIt != GEP->idx_end(); IdxIt++) {
+        Value *Idx = *IdxIt;
+        if (Instruction *NonConstIdx = dyn_cast<Instruction>(Idx)) {
+          if (LoopDependencies.count(NonConstIdx))
+            return true;
+        }
+      }
+    }
+  }
+  return false;
+}
+
+static void FindProblemBlocks(BasicBlock *Header, const SmallVectorImpl<BasicBlock *> &BlocksInLoop, SetVector<BasicBlock *> &ProblemBlocks) {
+  SetVector<BasicBlock *> BlocksInLoopSet(BlocksInLoop.begin(), BlocksInLoop.end());
+  SetVector<Instruction *> LoopDependencies;
+  SmallVector<PHINode *, 8> HeaderPHI;
+  for (Instruction &I : *Header) {
+    PHINode *PN = dyn_cast<PHINode>(&I);
+    if (!PN)
+      break;
+    FindAllDataDependency(PN, BlocksInLoopSet, LoopDependencies);
+  }
+
+  for (BasicBlock *BB : BlocksInLoop) {
+    if (DoesBlockNeedUnroll(BB, LoopDependencies))
+      ProblemBlocks.insert(BB);
+  }
+}
+
+static bool ContainsFloatingPointType(Type *Ty) {
+  if (Ty->isFloatingPointTy()) {
+    return true;
+  }
+  else if (Ty->isArrayTy()) {
+    return ContainsFloatingPointType(Ty->getArrayElementType());
+  }
+  else if (Ty->isVectorTy()) {
+    return ContainsFloatingPointType(Ty->getVectorElementType());
+  }
+  else if (Ty->isStructTy()) {
+    for (unsigned i = 0, NumStructElms = Ty->getStructNumElements(); i < NumStructElms; i++) {
+      if (ContainsFloatingPointType(Ty->getStructElementType(i)))
+        return true;
+    }
+  }
+  return false;
+}
+
+static bool Mem2Reg(Function &F, DominatorTree &DT, AssumptionCache &AC) {
+  BasicBlock &BB = F.getEntryBlock();  // Get the entry node for the function
+  bool Changed  = false;
+  std::vector<AllocaInst*> Allocas;
+  while (1) {
+    Allocas.clear();
+
+    // Find allocas that are safe to promote, by looking at all instructions in
+    // the entry node
+    for (BasicBlock::iterator I = BB.begin(), E = --BB.end(); I != E; ++I)
+      if (AllocaInst *AI = dyn_cast<AllocaInst>(I))       // Is it an alloca?
+        if (isAllocaPromotable(AI) &&
+          (!HLModule::HasPreciseAttributeWithMetadata(AI) || !ContainsFloatingPointType(AI->getAllocatedType())))
+          Allocas.push_back(AI);
+
+    if (Allocas.empty()) break;
+
+    PromoteMemToReg(Allocas, DT, nullptr, &AC);
+    Changed = true;
+  }
+
+  return Changed;
+}
+
+bool DxilLoopUnroll::runOnLoop(Loop *L, LPPassManager &LPM) {
```

Comment:
>runOnLoop [](start = 21, length = 9)

this function is really big, can't you extract some of the logic into smaller functions?